### PR TITLE
[core] Remove React.FC

### LINKED ISSUE #1325: [DataGrid] Use explicit definition instead of React.FC
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

We're typing many components with `React.FC`. One of the [problems](https://github.com/facebook/create-react-app/pull/8177) with this approach is that it provides an implicit definition of the `children` prop, even for those components that don't use it. We should favor an explicit definition instead.

```diff
diff --git a/packages/grid/_modules_/grid/components/GridRowCount.tsx b/packages/grid/_modules_/grid/components/GridRowCount.tsx
index 4d888ce..45ca4a3 100644
--- a/packages/grid/_modules_/grid/components/GridRowCount.tsx
+++ b/packages/grid/_modules_/grid/components/GridRowCount.tsx
@@ -1,7 +1,11 @@
 import * as React from 'react';
 import { GridApiContext } from './GridApiContext';

-export const GridRowCount: React.FC<{ rowCount: number }> = ({ rowCount }) => {
+export interface GridRowCountProps {
+  rowCount: number
+}
+
+export const GridRowCount = (props: GridRowCountProps) => {
+  const { rowCount } = props;
   const apiRef = React.useContext(GridApiContext);

   if (rowCount === 0) {
```

### PR DESCRIPTION
**Pull request by ZeeshanTamboli**
Resolves #1325

### CODE REVIEW

File: packages/grid/_modules_/grid/models/colDef/gridCheckboxSelection.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ export const gridCheckboxSelectionColDef: GridColDef = {
   disableClickEventBubbling: true,
   disableColumnMenu: true,
   valueGetter: (params) => params.api.getState().selection[params.id] !== undefined,
-  renderHeader: (params) => <GridHeaderCheckbox {...params} />,
+  renderHeader: () => <GridHeaderCheckbox />,
```
Comment (dtassone) [inline packages/grid/_modules_/grid/models/colDef/gridCheckboxSelection.tsx:20]: why do you need to remove the params here?

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/models/colDef/gridCheckboxSelection.tsx:20]: `GridHeaderCheckbox` component was not using any of the props (GridColumnHeaderParams) passed. So I think we need not type it,
```diff
- export const GridHeaderCheckbox: React.FC<GridColumnHeaderParams> = () =>
+ export const GridHeaderCheckbox = () => {
```
And also thus need to remove the `params` from here as well.

**Review by dtassone — APPROVED** [review]

File: packages/grid/_modules_/grid/components/GridStickyContainer.tsx [hunk 2]
```diff
@@ -1,16 +1,23 @@
...
+        minWidth: width,
+        maxWidth: width,
+        maxHeight: height,
+      }}
+    >
+      {children}
+    </div>
+  );
+};
 GridStickyContainer.displayName = 'GridStickyContainer';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/GridStickyContainer.tsx:23]: While we are at it, we could do
```suggestion
```

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/components/GridStickyContainer.tsx]: Updated

File: packages/grid/_modules_/grid/components/cell/GridCell.tsx [hunk 3]
```diff
@@ -35,9 +35,10 @@ export interface GridCellProps {
   value?: GridCellValue;
   width: number;
   cellMode?: GridCellMode;
+  children: React.ReactNode;
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/components/cell/GridCell.tsx:38]: It's a small thing, but we could narrow it a little more. `GridCell` always receives a `ReactElement` or null from `GridRowsCell`.

```suggestion
  children: React.ReactElement | null;
```

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/components/cell/GridCell.tsx]: Sure. Added

**Review by DanailH — APPROVED** [review]

**Pull request merged**